Phonopy: fixing atoms

tomdemeyere opened the pull request:
## Summary of Changes

I like the idea that you had, removing the parameter `fixed_atoms` and deal with constraints instead, some problems though:  users might want to optimise these atoms before hand during the `relax_job`, or sometimes not, how to deal with the constraints then? Also, this might be seen as a "hidden" feature?

This MR also **really** removes the second `get_phonopy` which somehow evaded both of our attention during the last PR 😂

### Checklist

- [ ] I have read the ["Guidelines" section](https://quantum-accelerators.github.io/quacc/dev/contributing.html#guidelines) of the contributing guide. Don't lie! 😉
- [ ] My PR is on a custom branch and is _not_ named `main`.
- [ ] I have added relevant, comprehensive [unit tests](https://quantum-accelerators.github.io/quacc/dev/contributing.html#unit-tests).

### Notes

- Your PR will likely not be merged without proper and thorough tests.
- If you are an external contributor, you will see a comment from [@buildbot-princeton](https://github.com/buildbot-princeton). This is solely for the maintainers.
- When your code is ready for review, ping one of the [active maintainers](https://quantum-accelerators.github.io/quacc/about/contributors.html#active-maintainers).

buildbot-princeton commented:
Can one of the admins verify this patch?

Andrew-S-Rosen commented:
> I like the idea that you had, removing the parameter fixed_atoms and deal with constraints instead, some problems though: users might want to optimise these atoms before hand during the relax_job, or sometimes not, how to deal with the constraints then? Also, this might be seen as a "hidden" feature?

This is worth thinking about. Let me think out loud here. 

In Scenario A, the user wants to relax all the atoms and then, afterwards, do a phonon calculation with a subset of atoms fixed. In this scenario, they should just run a `relax_job()` first and then pass the output to `phonon_flow()` with `run_relax=False`. 

```python
from quacc.recipes.emt.core import relax_job
from quacc.recipes.emt.phonons import phonon_flow

output = relax_job(atoms)
new_atoms = output["atoms"]
new_atoms.set_constraint(...)
output2 = phonon_flow(new_atoms, run_relax=False)
```

In Scenario B, the user wants to have a subset of atoms fixed during both the relax and the phonon flow. In this case, the only need to call `phonon_flow()` directly (with the `run_relax=True` default set) and their input `Atoms` objects with constraints.

So, in both scenarios, this is currently possible. While I agree that adding a dedicated keyword argument is more convenient for the user, the counterpoint to this is that it adds more lines to maintain even though it is not strictly necessary. Right now, virtually every recipe in quacc supports fixing atoms in any way you'd like simply by passing in an `Atoms` object with constraints pre-applied. We didn't have to do anything to achieve that; it comes "for free" by simply basing things around ASE. Admittedly, this could be communicated more clearly to end-users and should likely (at the very least) have a dedicated call out in the introductory section of the documentation: https://github.com/Quantum-Accelerators/quacc/issues/2074.

> This MR also really removes the second get_phonopy which somehow evaded both of our attention during the last PR 😂

Oops, yeah that was a mistake. I also like the addition you have made regarding symmetrization of the force constants.

tomdemeyere commented:
> the counterpoint to this is that it adds more lines to maintain even though it is not strictly necessary

To use ASE's constraints the only way that comes to my mind is to do:

``` python
fixed = []

for constr in atoms.constraints:
    if isinstance(constr, FixAtoms):
        fixed.extend(constr.get_indices())
```

Then remove the `fixed_atoms` parameter, but the rest of the code would not change.

Andrew-S-Rosen commented:
@tomdemeyere: Thanks. My point was more about the need for these changes at all if one can pass in an `Atoms` object with constraints already applied. It's not clear to me why we _need_ to have a `fixed_atoms` keyword argument when (as described in my prior comment) it is possible to use any recipe with any set of ASE constraints without having a dedicated keyword argument. Am I misunderstanding something about this?

The other changes, specifically regarding getting rid of the second `get_phonopy()` call and adding symmetrization of the force constants, are quite logical.

tomdemeyere commented:
Apologies for the lack of context. What we have to do to *effectively* fix some atoms when using phonopy is the following:

1. Construct the supercell but only sending the partial system (non-fixed) to phonopy, this way phonopy thinks the system being calculated is the non-fixed one. (lines 80 to 93 in quacc.atoms.phonons.get_phonopy)
2. Re-construct the full system manually by doing the same supercell transformation for the rest of the system (the fixed atoms) outside of phonopy (lines 95 to 99 in quacc.atoms.phonons.get_phonopy)
3. The generated displacements will only be for the non-fixed atoms due to point 1, but we still send the full systems to the calculator by adding all the fixed system to each displacement (lines 101 to 104 in quacc.recipes.common.phonons)
4. Phonopy expects forces with shape (len(disp), len(non_fixed_atoms), 3) so we manually extract these forces from the full calculations (lines 115 to 118 in quacc.recipes.common.phonons)
5. Post-processing as usual... This whole process drastically reduces the number of displacement required, for an adsorbate on a slab for example you might end up calculating the displacement of the adsorbate only. As you might guess this is a huge approximation because you assume that modes from the fixed system do not interact with the ones of the non-fixed system.
6. As a consequence, the symmetry of the force constant is broken, which we manually try to enforce later.

ASE's constraints unfortunately are not of help here. This whole trick have to be done manually

Andrew-S-Rosen commented on src/quacc/atoms/phonons.py:
```diff
@@ -62,24 +66,38 @@ def get_phonopy(
-    return phonon
+
+    if len(fixed) > 0:
+        fixed = get_phonopy_structure(AseAtomsAdaptor.get_structure(fixed))
+        fixed = phonopy.structure.cells.get_supercell(fixed, supercell_matrix)
+        fixed = AseAtomsAdaptor.get_atoms(get_pmg_structure(fixed))
```
These are three different types of objects yet they all have the same variable name; a bit confusing. This should be clarified.

Andrew-S-Rosen commented on src/quacc/recipes/common/phonons.py:
```diff
@@ -79,17 +84,23 @@ def phonon_subflow(
         phonopy_kwargs=phonopy_kwargs,
     )
+
+    fixed_atoms = np.full(len(phonon.supercell), False)
```
Please rename as `fixed_indices` instead of `fixed_atoms`.

Andrew-S-Rosen commented on src/quacc/recipes/common/phonons.py:
```diff
@@ -79,17 +84,23 @@ def phonon_subflow(
     )
+
+    fixed_atoms = np.full(len(phonon.supercell), False)
+    fixed_atoms = np.append(fixed_atoms, [True] * len(atoms_to_add))
+    fixed_atoms = fixed_atoms.astype(bool)
+
     supercells = [
-        phonopy_atoms_to_ase_atoms(s) for s in phonon.supercells_with_displacements
+        phonopy_atoms_to_ase_atoms(s) + atoms_to_add
+        for s in phonon.supercells_with_displacements
     ]
```
I think I need to return to this later to fully digest what's going on, but I'll trust you on it for the time being.

Andrew-S-Rosen reviewed (commented):
@tomdemeyere: Thank you very much for your helpful reply. I had (incorrectly) assumed based on your PR description that this was mainly centered around ease-of-use. However, I now understand that it is actually _correcting_ the behavior for fixed atoms and that simply supplying the `Atoms` object with `FixAtoms` constraints is not sufficient. I am familiar with the typical assumption about decoupling of vibrational modes between surface and adsorbate but have normally used ASE's `Vibrations` module and ASE's thermo modules for this instead of phonon calculations via phonopy. If I recall correctly, ASE's `Vibrations` module is smart enough not to displace and vibrate atoms that are fixed.

Anyway, I have left several comments below. Apologies in advance about some of them being a bit pedantic, but it will help increase the clarity for future me.

tomdemeyere commented on src/quacc/atoms/phonons.py:
```diff
@@ -62,24 +66,38 @@ def get_phonopy(
     if supercell_matrix is None and min_lengths is not None:
         n_supercells = np.round(np.ceil(min_lengths / atoms.cell.lengths()))
-        supercell_matrix = np.diag([n_supercells, n_supercells, n_supercells])
+        supercell_matrix = np.diag(np.diag([n_supercells, n_supercells, n_supercells]))
```
[n_supercells, n_supercells, n_supercells] is a 2D array, applying np.diag will make it 1D, applying two times will make it 2D, is there something I am missing? If you want a 2D array you should just keep np.array([n_supercells, n_supercells, n_supercells])?

Andrew-S-Rosen replied:
Phonopy can accept a supercell matrix that is 3x3. Doing `np.diag([1, 2, 3])` makes a 3x3 matrix. However, doing `np.diag(np.diag([1, 2, 3]))` makes a 1D array of dimensions `(3,)`. No?

Andrew-S-Rosen replied:
See https://github.com/phonopy/phonopy/blob/430bb717fd1af28515bfbd776986202c9e73bf35/doc/phonopy-module.md?plain=1#L46-L48 and the examples that follow.

> The first step is to create a `Phonopy` object with at least two arguments, a
unit cell (`PhonopyAtoms` object, see {ref}`phonopy_Atoms`) and a supercell
matrix (3x3 array, see {ref}`variable_supercell_matrix`).

Andrew-S-Rosen replied:
Never mind. I see the problem...

tomdemeyere replied:
`supercell_matrix = np.eye(3)* np.round(np.ceil(min_lengths / atoms.cell.lengths()))` should do the trick

tomdemeyere replied:
Or just doing np.diag, which is safer if somehow cell is not 3D

Andrew-S-Rosen replied:
Yup! Either approach works. 👍

tomdemeyere commented:
I don't really like the fact that `get_phonopy` now returns a tuple but I don't see any other non-convoluted way to do that :/

Andrew-S-Rosen commented:
> I don't really like the fact that `get_phonopy` now returns a tuple but I don't see any other non-convoluted way to do that :/

I agree that it feels a bit wrong since the `Atoms` object isn't needed in most cases and the function is primarily about getting a `Phonopy` object. If you end up thinking of a better alternative, by all means go ahead. If not though, I won't let it block the PR.

Andrew-S-Rosen commented on tests/core/recipes/emt_recipes/test_emt_phonons.py:
```diff
@@ -71,3 +72,33 @@ def test_phonon_flow_v4(tmp_path, monkeypatch):
+    output_fixed = phonon_flow(full_system, run_relax=True, min_lengths=5.0)
+
+    # Should be very close but not exactly the same
+    assert output["results"]["mesh_properties"]["frequencies"] == pytest.approx(
+        output_fixed["results"]["mesh_properties"]["frequencies"], rel=0.0, abs=1e-5
+    )
```
@tomdemeyere: Is this PR good to go from your end?

Also, could you please explain this test for me? Why would the frequencies be the same if two of the atoms are fixed?

tomdemeyere replied:
The two molecules are far away, if you fix one of them it should not affect the other.

In the last one we fix one atom in each molecule giving completely wrong results...

The PR is good to go, let me keep in mind the get_phonopy problem I will come back to it if I find a solution

tomdemeyere commented:
Last minute thought:

"Fixing" atoms during phonons calculations is not a real thing and many people would simply call it blasphemy I think it might be wise to add an extra warning with the logger, especially considering that many Phonopy will not expect this feature to exist, what do you think?

Andrew-S-Rosen commented:
@tomdemeyere: Thanks for the suggestion. I have been gravitating back towards having the `fix_indices` keyword argument for this reason; the user really has to directly request it and know what they are getting themselves into. I have this taken care of in a custom branch I am working on right now (not public yet).

My main hesitations right now are that `quacc.atoms.phonons.get_phonopy()` is just super messy/complicated. There is a ton of back-and-forth converting happening, and it is not trivial to follow along. I am trying to make some headway in cleaning that up, although I don't know if I'll be successful... in any case, I'll post here later this afternoon whether I'm successful or not.

tomdemeyere commented:
"I have been gravitating back towards having the fix_indices keyword argument for this reason; the user really has to directly request it and know what they are getting themselves into."

I think this might be better. When you suggested to use ASE's constraints I got hooked as it is definitely more consistent and "appealing", but after thinking about it a little bit more I would vote for keeping `fix_indices` for the reason you invoked.

About Phonopy in Quacc: It might be more interesting to make simpler Phonopy functions instead of whole `flow`s, aiming for increased modularity, even at the cost of increased complexity. Phonopy is a very complex package and in the current state available `flow`s can only scratch the package's surface due to the lack of possible parameters customization. This might also feel better in a workflow sense(from what I understand): if you would be better off resolving in the middle of a flow something is off (you know better, just shooting an idea here)

Users seeking something simpler might be better off being redirected to ASE's phonon/vibrations workflows (which are still powerfuls but arguably less complex).

Andrew-S-Rosen commented:
@tomdemeyere: Apologies again that this hasn't been as trivial to merge as I initially believed. 

I have opened up PR #2108 to demonstrate some areas for improvement. Namely, I want to draw your attention to [`quacc.atoms.phonons`](https://github.com/Quantum-Accelerators/quacc/blob/8da3bed807c9ba5355671e1ac5eae75136d6a539/src/quacc/atoms/phonons.py). Note that by sticking with the pymatgen `Structure` object, we can avoid back-and-forths between different atoms representations. Additionally, throughout this PR there were several areas where clearer variable names were needed; it got very hectic.

There are still some problems with PR #2108, however:
- The tests are failing, so I clearly made a mistake somewhere. But I don't have time to dig into this further.
- The `quacc.recipes.common.phonons.phonon_subflow` is getting increasingly complicated with this fix atoms behavior. While not a dealbreaker, it is a bit unfortunate.
- The fact that `quacc.atoms.phonons.get_phonopy` returns a `tuple` even when `fixed_indices=None` is a bit annoying, although it is less severe of a problem in #2108 than here in #2073.

Feel free to copy or pick things up from there. Unfortunately, I don't feel comfortable merging #2073 until this gets streamlined, especially for what amounts to a crude approximation in many cases (although one that is quite common).

tomdemeyere commented:
I think I just got an idea that might avoid quite a lot of the current mess to fix atoms, Also avoiding get_phonopy returning a tuple.

Let me have a look tomorrow.

Andrew-S-Rosen commented:
Closing in favor of #2110.
